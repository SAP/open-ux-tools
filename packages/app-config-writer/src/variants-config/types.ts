type UI5FlexLayer = 'VENDOR' | 'CUSTOMER_BASE';
interface RtaConfig {
    layer: UI5FlexLayer;
    options?: {
        [key: string]: unknown;
        baseId?: string;
        projectId?: string;
        scenario?: string;
        appName?: string;
    };
    editors: any[];
}
interface Intent {
    object: string;
    action: string;
}
interface App {
    target: string;
    local?: string;
    componentId?: string;
    intent?: Intent;
}
interface FlpConfig {
    path: string;
    intent: Intent;
    libs?: boolean;
    apps: App[];
    theme?: string;
    init?: string;
}
interface OptionalTestConfig {
    path: string;
    init: string;
    pattern: string;
}
interface TestConfig extends Partial<OptionalTestConfig> {
    framework: 'OPA5' | 'QUnit' | 'Testsuite';
}
interface AdpPreviewConfig {
    target: any;
    ignoreCertErrors?: boolean;
}
export interface MiddlewareConfig {
    flp?: Partial<FlpConfig>;
    test?: TestConfig[];
    rta?: RtaConfig;
    adp?: AdpPreviewConfig;
    debug?: boolean;
}
export interface DeprecatedConfig {
    component: string;
    libs?: boolean;
    ui5Theme?: string;
}

export type PreviewConfigOptions = DeprecatedConfig | MiddlewareConfig;
