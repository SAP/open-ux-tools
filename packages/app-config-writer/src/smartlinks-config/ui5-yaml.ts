import type { Editor } from 'mem-fs-editor';
import { join } from 'path';
import type { ToolsLogger } from '@sap-ux/logger';
import { FileName, readUi5Yaml } from '@sap-ux/project-access';
import type {
    AbapDeployConfig,
    CustomMiddleware,
    FioriToolsProxyConfig,
    FioriToolsServeStaticConfig,
    FioriToolsServeStaticPath,
    UI5Config
} from '@sap-ux/ui5-config';
import { t } from '../i18n';
import type { TargetConfig } from '../types';
import { DeployConfig } from '../types';

/**
 * Reads and returns target information from ui5-deploy.yaml, if existing.
 *
 * @param basePath - path to project root, where ui5-deploy.yaml is
 * @returns {TargetConfig} target definition for deploy configuration
 */
export async function readUi5DeployConfigTarget(basePath: string): Promise<TargetConfig> {
    const ui5DeployYaml = await readUi5Yaml(basePath, FileName.UI5DeployYaml);
    const customTask = ui5DeployYaml.findCustomTask<AbapDeployConfig>(DeployConfig.DeployToAbap);
    if (!customTask?.configuration?.target) {
        throw Error(t('error.noTarget', { file: `(${FileName.UI5DeployYaml})` }));
    }
    const { target, ignoreCertError } = customTask?.configuration;
    return { target, ignoreCertErrors: ignoreCertError };
}

/**
 * Checks if 'fiori-tools-servestatic' configuration is already existing.
 *
 * @param existingPaths existing configuration paths under 'fiori-tools-servestatic'
 * @param existingFioriToolsProxy - flag if fiori tools proxy middleware is already existing
 * @returns 'fiori-tools-servestatic' configuration, if not existing yet
 */
const getFioriToolsServeStaticMiddlewareConfig = (
    existingPaths: FioriToolsServeStaticPath[],
    existingFioriToolsProxy: boolean
): CustomMiddleware<FioriToolsServeStaticConfig> | undefined => {
    const configPath: FioriToolsServeStaticPath = { path: '/appconfig', src: 'appconfig', fallthrough: false };
    if (existingPaths.find((existing) => existing.path === configPath.path && existing.src === configPath.src)) {
        return undefined;
    }
    const paths = [...existingPaths, configPath];
    const beforeAfterMiddleware = existingFioriToolsProxy
        ? { beforeMiddleware: 'fiori-tools-proxy' }
        : { afterMiddleware: 'compression' };
    return {
        name: 'fiori-tools-servestatic',
        ...beforeAfterMiddleware,
        configuration: { paths }
    };
};

/**
 * Reads and adds servestatic configuration to ui5/-local/-mock.yaml files.
 *
 * @param basePath - path to project root, where ui5.yaml is
 * @param fs - mem-fs reference to be used for file access
 * @param logger - logger
 */
export async function addUi5YamlServeStaticMiddleware(
    basePath: string,
    fs: Editor,
    logger?: ToolsLogger
): Promise<void> {
    const ui5Yamls = [FileName.Ui5Yaml, FileName.Ui5MockYaml, FileName.Ui5LocalYaml];
    for (const ui5Yaml of ui5Yamls) {
        let ui5YamlConfig: UI5Config;
        try {
            ui5YamlConfig = await readUi5Yaml(basePath, ui5Yaml);
        } catch (error) {
            logger?.debug(`File ${ui5Yaml} not existing`);
            continue;
        }
        const existingFioriToolsProxy = !!ui5YamlConfig.findCustomMiddleware<FioriToolsProxyConfig>(
            DeployConfig.FioriToolsProxy
        );
        const appServeStaticMiddleware = ui5YamlConfig.findCustomMiddleware<FioriToolsServeStaticConfig>(
            DeployConfig.FioriToolsServestatic
        );
        const middleware = getFioriToolsServeStaticMiddlewareConfig(
            appServeStaticMiddleware?.configuration.paths || [],
            existingFioriToolsProxy
        );
        if (middleware) {
            const yamlConfig = ui5YamlConfig.updateCustomMiddleware(middleware);
            const yaml = yamlConfig.toString();
            fs.write(join(basePath, ui5Yaml), yaml);
        }
    }
}
