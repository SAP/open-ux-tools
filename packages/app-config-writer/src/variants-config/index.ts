export * from './generateVariantsConfig';
// export { getTargetDefinition, getLocalStoredCredentials } from './utils';
