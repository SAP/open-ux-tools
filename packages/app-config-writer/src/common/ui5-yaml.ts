import { basename, join } from 'path';
import { MiddlewareConfigs } from '../types';
import { FileName, type Package, readUi5Yaml } from '@sap-ux/project-access';
import type { Editor } from 'mem-fs-editor';
import type { ToolsLogger } from '@sap-ux/logger';
import type { PreviewConfigOptions } from '../types';
import type { CustomMiddleware, FioriAppReloadConfig, UI5Config } from '@sap-ux/ui5-config';
import { getPreviewMiddleware, isFioriToolsDeprecatedPreviewConfig } from './utils';
import type { MiddlewareConfig as PreviewConfig } from '@sap-ux/preview-middleware';
import { updatePreviewMiddlewareConfig } from '../preview-config/ui5-yaml';
import { getRunScriptForYamlConfig } from './package-json';

/**
 * Gets the reload middleware form the provided yamlConfig.
 * The middleware can either be named 'fiori-tools-appreload' or 'reload-middleware'.
 * If the middleware is found, a delay of 300ms will be inserted.
 *
 * @param yamlConfig - the yaml configuration to use
 * @returns reload middleware configuration if found or undefined
 */
export async function getEnhancedReloadMiddleware(
    yamlConfig: UI5Config
): Promise<CustomMiddleware<FioriAppReloadConfig> | undefined> {
    const reloadMiddleware =
        yamlConfig.findCustomMiddleware<FioriAppReloadConfig>(MiddlewareConfigs.FioriToolsAppreload) ??
        yamlConfig.findCustomMiddleware<FioriAppReloadConfig>(MiddlewareConfigs.ReloadMiddleware);
    if (!reloadMiddleware) {
        return undefined;
    }
    if (!reloadMiddleware?.configuration?.delay) {
        reloadMiddleware.configuration = { ...reloadMiddleware.configuration, delay: 300 };
    }
    return reloadMiddleware;
}

/**
 * Creates a preview middleware configuration based on the presence of the @sap/ux-ui5-tooling dependency.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param basePath - path to project root, where package.json and ui5.yaml is
 * @returns 'fiori-tools-preview' or 'preview-middleware' configuration
 */
export function createPreviewMiddlewareConfig(fs: Editor, basePath: string): CustomMiddleware<PreviewConfigOptions> {
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    return {
        name: packageJson?.devDependencies?.['@sap/ux-ui5-tooling']
            ? MiddlewareConfigs.FioriToolsPreview
            : MiddlewareConfigs.PreviewMiddleware,
        afterMiddleware: 'compression'
    } as CustomMiddleware<PreviewConfigOptions>;
}

/**
 * Checks the project for ui5 configuration yaml files and reads the configuration to update the preview and reload middlewares.
 * If a reload middleware exists, then a delay of 300ms will be inserted and the preview middleware will be set afterward.
 *
 * @param fs - mem-fs reference to be used for file access
 * @param basePath - path to project root, where package.json and ui5.yaml is
 * @param yamlPath - path to the ui5*.yaml file passed by cli
 * @param logger - logger
 */
export async function updateMiddlewaresForPreview(
    fs: Editor,
    basePath: string,
    yamlPath?: string,
    logger?: ToolsLogger
): Promise<void> {
    const ui5YamlFile = yamlPath ? basename(yamlPath) : FileName.Ui5Yaml;
    const ui5YamlConfig = await readUi5Yaml(basePath, ui5YamlFile, fs);

    let previewMiddleware = await getPreviewMiddleware(ui5YamlConfig);
    if (!previewMiddleware) {
        logger?.warn(`No preview middleware found in ${ui5YamlFile}. Preview middleware will be added.`);
        previewMiddleware = createPreviewMiddlewareConfig(fs, basePath);
    } else {
        const script = getRunScriptForYamlConfig(ui5YamlFile, fs, basePath);
        if (script) {
            previewMiddleware = await updatePreviewMiddlewareConfig(previewMiddleware, script, basePath, fs, logger);
        } else {
            //if we don't find a script for flp.path and intent we assume the default values and sanitize the config
            previewMiddleware = sanitizePreviewMiddleware(previewMiddleware) as CustomMiddleware<PreviewConfig>;
        }
    }
    const reloadMiddleware = await getEnhancedReloadMiddleware(ui5YamlConfig);
    if (reloadMiddleware) {
        previewMiddleware.afterMiddleware = reloadMiddleware.name;
        ui5YamlConfig.updateCustomMiddleware(reloadMiddleware);
        logger?.debug(`Updated reload middleware in ${ui5YamlFile}.`);
    }

    ui5YamlConfig.updateCustomMiddleware(previewMiddleware);
    fs.write(join(basePath, ui5YamlFile), ui5YamlConfig.toString());
    logger?.debug(`Updated preview middleware in ${ui5YamlFile}.`);
}

/**
 * Sanitizes the preview middleware configuration.
 *
 * In case of an outdated preview configuration, the following changes will be applied:
 * - property 'ui5Theme' will be moved to 'flp.theme'.
 * - no longer used property 'component' will be removed.
 *
 * @param previewMiddleware - the preview middleware
 * @returns the sanitized preview middleware
 */
export function sanitizePreviewMiddleware(
    previewMiddleware: CustomMiddleware<PreviewConfigOptions>
): CustomMiddleware<PreviewConfig | undefined> {
    if (!isFioriToolsDeprecatedPreviewConfig(previewMiddleware.configuration)) {
        return previewMiddleware as CustomMiddleware<PreviewConfig>;
    }
    const ui5Theme = previewMiddleware.configuration.ui5Theme;
    delete (previewMiddleware as CustomMiddleware<PreviewConfig | undefined>).configuration;

    if (!ui5Theme) {
        return previewMiddleware as unknown as CustomMiddleware<undefined>;
    }

    const configuration = {} as PreviewConfig;
    configuration.flp = {};
    configuration.flp.theme = ui5Theme;
    previewMiddleware.configuration = configuration;
    return previewMiddleware as CustomMiddleware<PreviewConfig>;
}
