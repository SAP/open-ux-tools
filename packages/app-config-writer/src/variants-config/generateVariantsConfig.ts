import { create } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { addPreviewMiddlewareToYaml } from './ui5-yaml';
import { addVariantsManagementScript } from './package-json';
import type { Editor } from 'mem-fs-editor';
import type { ToolsLogger } from '@sap-ux/logger';

/**
 * Add variants configuration to an app or project.
 *
 * @param basePath - the base path where the package.json and ui5.yaml is
 * @param logger - logger
 * @param fs - the memfs editor instance
 * @returns Promise<Editor> - memfs editor instance with updated files
 */
export async function generateVariantsConfig(basePath: string, logger?: ToolsLogger, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    await addVariantsManagementScript(fs, basePath, logger);
    await addPreviewMiddlewareToYaml(fs, basePath, logger);
    return fs;
}
