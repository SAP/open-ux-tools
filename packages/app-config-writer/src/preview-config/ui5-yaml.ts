import { basename, join } from 'path';
import { createPreviewMiddlewareConfig } from '../variants-config/ui5-yaml';
import { ensurePreviewMiddlewareDependency, extractUrlDetails, isValidPreviewScript } from './package-json';
import { FileName, getAllUi5YamlFileNames, getWebappPath, readUi5Yaml, type Package } from '@sap-ux/project-access';
import { getPreviewMiddleware, isFioriToolsDeprecatedPreviewConfig } from '../variants-config/utils';
import { renameSandbox } from './preview-files';
import type { CustomMiddleware, UI5Config } from '@sap-ux/ui5-config';
import type { Editor } from 'mem-fs-editor';
import type {
    FlpConfig,
    MiddlewareConfig as PreviewConfig,
    TestConfig,
    DefaultFlpPath,
    DefaultIntent
} from '@sap-ux/preview-middleware';
import type { PreviewConfigOptions } from '../types';
import type { ToolsLogger } from '@sap-ux/logger';

const DEFAULT_FLP_PATH: DefaultFlpPath = '/test/flp.html';

const DEFAULT_INTENT: DefaultIntent = {
    object: 'app',
    action: 'preview'
};

const DEFAULT_TEST_CONFIGS: TestConfig[] = [
    { framework: 'Testsuite', path: 'test/testsuite.qunit.html' },
    { framework: 'OPA5', path: 'test/opaTests.qunit.html' },
    { framework: 'QUnit', path: 'test/unitTests.qunit.html' }
];

/**
 * Checks if a script can be converted based on the used UI5 yaml configuration file.
 *
 * A script can be converted when the used UI5 yaml configuration file exists and complies with the schema validation.
 *
 * @param ui5Yaml - the name of the UI5 yaml configuration file
 * @param scriptName - the name of the script from the package.json file
 * @param ui5YamlFileNames - the UI5 yaml configuration file names
 * @param logger logger to report info to the user
 * @returns indicator if the script is to be converted
 */
function isUi5YamlToBeConverted(
    ui5Yaml: string,
    scriptName: string,
    ui5YamlFileNames: string[],
    logger?: ToolsLogger
): boolean {
    if (!ui5YamlFileNames.includes(ui5Yaml)) {
        logger?.warn(
            `Skipping script '${scriptName}', because the UI5 YAML configuration file, '${ui5Yaml}', could not be found.`
        );
        return false;
    }
    return true;
}

/**
 * Check if a UI5 yaml configuration file has already been converted based on another script.
 *
 * This is done to avoid overwriting of already converted UI5 yaml configurations.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param ui5Yaml - the name of the UI5 yaml configuration file
 * @param scriptName - the name of the script from the package.json file
 * @param script - the content of the script from the package.json file
 * @param convertTests - indicator if test suite and test runner should be included in the conversion
 * @param logger logger to report info to the user
 * @returns indicator if the UI5 yaml configuration file has already been converted
 */
async function isUi5YamlFlpPathAlreadyConverted(
    fs: Editor,
    basePath: string,
    ui5Yaml: string,
    scriptName: string,
    script: string,
    convertTests: boolean,
    logger?: ToolsLogger
): Promise<boolean> {
    const yamlConfigAlreadyAdjusted =
        Object.keys(
            fs.dump(basePath, (file) => {
                return file.basename === ui5Yaml && file.state === 'modified';
            })
        ).length > 0;
    const flpPath = ((await getPreviewMiddleware(undefined, basePath, ui5Yaml, fs)) as CustomMiddleware<PreviewConfig>)
        ?.configuration?.flp?.path;
    const { path: scriptPath } = extractUrlDetails(script);
    if (yamlConfigAlreadyAdjusted && flpPath != scriptPath && (convertTests ? !isTestPath(scriptPath) : true)) {
        logger?.warn(
            `Skipping script'${scriptName}', because another script also refers to UI5 YAML configuration file, '${ui5Yaml}'. Adjust the 'flp.path' property in the UI5 YAML configuration file to the correct endpoint or create a separate UI5 YAML configuration file for script '${scriptName}'. ${ui5Yaml} currently uses ${
                flpPath ?? DEFAULT_FLP_PATH
            } whereas script '${scriptName}' uses '${scriptPath}'.`
        );
        return true;
    }
    return false;
}

/**
 * Checks if the passed path is a FLP path.
 *
 * @param path - the path
 * @param configuration - the preview configuration
 * @returns indicator if the path is an FLP path
 */
function isFlpPath(path: string | undefined, configuration: PreviewConfig): boolean {
    if (!path) {
        return false;
    }
    const isRtaEditorPath = configuration.rta?.editors?.some((editor) => editor.path === path) ?? false;
    return !isRtaEditorPath && !isTestPath(path, configuration);
}

/**
 * Check if the path is a test path.
 * 1) path matches pattern '**.qunit.html'
 * 2) path is being used as test configuration path in yaml configuration.
 *
 * @param path - the path
 * @param configuration - the preview configuration
 * @returns indicator if the path is a test path
 */
export function isTestPath(path: string | undefined, configuration?: PreviewConfig): boolean {
    if (!path) {
        return false;
    }
    //todo: how to ensure we don't mistake other test scripts for preview scripts?
    if (path.includes('.qunit.html')) {
        return true;
    }
    return configuration?.test?.some((testConfig) => testConfig.path === path) ?? false;
}

/**
 * Sanitizes the preview middleware configuration.
 *
 * In case of an outdated preview configuration, the following changes will be applied:
 * - property 'ui5Theme' will be moved to 'flp.theme'.
 * - no longer used property 'component' will be removed.
 *
 * @param previewMiddleware - the preview middleware
 * @returns the sanitized preview middleware
 */
function sanitizePreviewMiddleware(
    previewMiddleware: CustomMiddleware<PreviewConfigOptions>
): CustomMiddleware<PreviewConfig | undefined> {
    if (!isFioriToolsDeprecatedPreviewConfig(previewMiddleware.configuration)) {
        return previewMiddleware as CustomMiddleware<PreviewConfig>;
    }
    const ui5Theme = previewMiddleware.configuration.ui5Theme;
    delete (previewMiddleware as CustomMiddleware<PreviewConfig | undefined>).configuration;

    if (!ui5Theme) {
        return previewMiddleware as unknown as CustomMiddleware<undefined>;
    }

    const configuration = {} as PreviewConfig;
    configuration.flp = {};
    configuration.flp.theme = ui5Theme;
    previewMiddleware.configuration = configuration;
    return previewMiddleware as CustomMiddleware<PreviewConfig>;
}

/**
 * Extract the UI5 yaml configuration file name from the script.
 *
 * @param script - the content of the script from the package.json file
 * @returns the UI5 yaml configuration file name or 'ui5.yaml' as default
 */
export function extractYamlConfigFileName(script: string): string {
    return / (?:--config|-c) (\S*)/.exec(script)?.[1] ?? FileName.Ui5Yaml;
}

/**
 * Processes the passed UI5 yaml configuration file and reads out the preview middleware configuration.
 *
 * If the preview middleware creation is not skipped and none is given, one will be created.
 * The configuration for the preview middleware creation will be based on the parameters of the passed script.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param ui5Yaml - the name of the UI5 yaml configuration file
 * @param script - the content of the script
 * @param skipPreviewMiddlewareCreation - (default: false) indicator if the preview middleware creation should be skipped if no preview middleware is configured.
 */
export async function processUi5YamlConfig(
    fs: Editor,
    basePath: string,
    ui5Yaml: string,
    script: string,
    skipPreviewMiddlewareCreation = false
): Promise<void> {
    let ui5YamlConfig: UI5Config;
    try {
        ui5YamlConfig = await readUi5Yaml(basePath, ui5Yaml, fs, { validateSchema: true });
    } catch (error) {
        throw new Error(`An error occurred when reading '${ui5Yaml}': ${error.message}`);
    }
    let previewMiddleware = await getPreviewMiddleware(ui5YamlConfig);

    if (skipPreviewMiddlewareCreation && !previewMiddleware) {
        return;
    }
    if (!previewMiddleware) {
        previewMiddleware = createPreviewMiddlewareConfig(fs, basePath);
    }

    const { path, intent } = extractUrlDetails(script);
    previewMiddleware = updatePreviewMiddlewareConfig(previewMiddleware, intent, path);

    ui5YamlConfig.updateCustomMiddleware(previewMiddleware);
    const yamlPath = join(basePath, ui5Yaml);
    fs.write(yamlPath, ui5YamlConfig.toString());
}

/**
 * Creates a preview middleware configuration.
 *
 * It will sanitize the given preview middleware configuration and construct the flp configuration out of the given intent and path.
 *
 * @param previewMiddleware - the preview middleware configuration
 * @param intent - the intent
 * @param path - the flp path
 * @returns the preview middleware configuration
 */
export function updatePreviewMiddlewareConfig(
    previewMiddleware: CustomMiddleware<PreviewConfigOptions>,
    intent: FlpConfig['intent'] | undefined,
    path: string | undefined
): CustomMiddleware<PreviewConfigOptions> {
    const defaultIntent = `${DEFAULT_INTENT.object}-${DEFAULT_INTENT.action}`;
    const newMiddlewareConfig = sanitizePreviewMiddleware(previewMiddleware);

    //copy of configuration to avoid ending up with an empty configuration object in some cases
    const configuration = { ...newMiddlewareConfig.configuration };

    let writeConfig = false;
    if (isFlpPath(path, configuration)) {
        //adjust path but respect defaults
        if (!path?.includes(DEFAULT_FLP_PATH)) {
            configuration.flp = configuration.flp ?? {};
            configuration.flp.path = path;
            writeConfig = true;
        }
        //adjust intent but respect defaults
        if (intent && `${intent?.object}-${intent?.action}` !== defaultIntent) {
            configuration.flp = configuration.flp ?? {};
            configuration.flp.intent = {
                object: intent.object,
                action: intent.action
            };
            writeConfig = true;
        }
    } else if (isTestPath(path, configuration)) {
        configuration.test = updateTestConfig(configuration.test, path);
        writeConfig = true;
    }

    if (writeConfig) {
        newMiddlewareConfig.configuration = configuration;
    }

    return newMiddlewareConfig as CustomMiddleware<PreviewConfig>;
}

/**
 * Update the test configuration.
 *
 * @param testConfiguration - the test configuration
 * @param path - the path
 * @returns the updated test configuration
 */
export function updateTestConfig(
    testConfiguration: PreviewConfig['test'],
    path: string | undefined
): PreviewConfig['test'] {
    testConfiguration = testConfiguration ?? [];

    let framework: TestConfig['framework'] | undefined;
    if (path?.includes('testsuite.qunit.html')) {
        framework = 'Testsuite';
    } else if (path?.includes('opaTests.qunit.html')) {
        framework = 'OPA5';
    } else if (path?.includes('unitTests.qunit.html')) {
        framework = 'QUnit';
    }

    if (!framework) {
        return testConfiguration;
    }

    const defaultPath = DEFAULT_TEST_CONFIGS.find((config) => config.framework === framework)?.path ?? '';
    const testConfig = testConfiguration.find((test) => test.framework === framework);
    if (testConfig) {
        testConfig.path = path;
        if (testConfig.path === defaultPath) {
            //sanitize default path
            delete testConfig.path;
        }
    } else if (path?.includes(defaultPath)) {
        testConfiguration.push({ framework });
    } else {
        testConfiguration.push({ framework, path });
    }
    return testConfiguration;
}

/**
 * Updates the default test configurations in the 'ui5.yaml' in case no test config exists in any UI5 configuration file.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param logger logger to report info to the user
 */
export async function updateDefaultTestConfig(fs: Editor, basePath: string, logger?: ToolsLogger): Promise<void> {
    const ui5YamlFileNames = await getAllUi5YamlFileNames(basePath, fs);
    for (const ui5Yaml of ui5YamlFileNames.filter((ui5Yaml) => ui5Yaml !== FileName.Ui5Yaml)) {
        const ui5YamlConfig = await readUi5Yaml(basePath, ui5Yaml, fs);
        const previewMiddleware = (await getPreviewMiddleware(ui5YamlConfig)) as CustomMiddleware<PreviewConfig>;
        if (previewMiddleware.configuration.test) {
            return;
        }
    }
    let ui5YamlConfig: UI5Config;
    try {
        ui5YamlConfig = await readUi5Yaml(basePath, FileName.Ui5Yaml, fs);
    } catch (error) {
        logger?.info(
            `The UI5 YAML configuration file 'ui5.yaml', can't be updated according to support test frameworks: '${error}'. Please manually add the test configuration to the UI5 YAML configuration file used for testing according to https://github.com/SAP/open-ux-tools/tree/main/packages/preview-middleware#configuration-option-test.`
        );
        return;
    }
    const previewMiddleware = (await getPreviewMiddleware(ui5YamlConfig)) as CustomMiddleware<PreviewConfig>;
    DEFAULT_TEST_CONFIGS.forEach((defaultTest) => {
        if (
            previewMiddleware.configuration?.test?.some((testConfig) => testConfig.framework === defaultTest.framework)
        ) {
            //do not touch existing test config
            return;
        }
        previewMiddleware.configuration.test = updateTestConfig(previewMiddleware.configuration.test, defaultTest.path);
        logger?.info(
            `The UI5 YAML configuration file 'ui5.yaml', has been updated to support the test framework '${defaultTest.framework}'. Please consider transferring the test configuration to the UI5 YAML configuration file used for testing.`
        );
    });
    ui5YamlConfig.updateCustomMiddleware(previewMiddleware);
    const yamlPath = join(basePath, FileName.Ui5Yaml);
    fs.write(yamlPath, ui5YamlConfig.toString());
}

/**
 * Updates the preview middleware configurations according to the scripts they are being used in package.json.
 *
 * It will process all given UI5 configuration yaml files and check if the preview middleware configuration must be updated based on a given script.
 * If a script is valid, the preview middleware configuration will be updated and the corresponding file renamed.
 * UI5 configuration yaml files which are not used in any script will be sanitized.
 *
 * @param fs - file system reference
 * @param basePath - base path to be used for the conversion
 * @param convertTests - indicator if test suite and test runner should be included in the conversion
 * @param logger logger to report info to the user
 */
export async function updatePreviewMiddlewareConfigs(
    fs: Editor,
    basePath: string,
    convertTests: boolean,
    logger?: ToolsLogger
): Promise<void> {
    const ui5YamlFileNames = await getAllUi5YamlFileNames(basePath, fs);
    const unprocessedUi5YamlFileNames = [...ui5YamlFileNames];
    const packageJsonPath = join(basePath, 'package.json');
    const packageJson = fs.readJSON(packageJsonPath) as Package | undefined;
    for (const [scriptName, script] of Object.entries(packageJson?.scripts ?? {})) {
        if (!script || !isValidPreviewScript(scriptName, script, convertTests)) {
            continue;
        }

        const ui5Yaml = basename(extractYamlConfigFileName(script));
        unprocessedUi5YamlFileNames.splice(unprocessedUi5YamlFileNames.indexOf(ui5Yaml), 1);

        if (
            !isUi5YamlToBeConverted(ui5Yaml, scriptName, ui5YamlFileNames, logger) ||
            (await isUi5YamlFlpPathAlreadyConverted(fs, basePath, ui5Yaml, scriptName, script, convertTests, logger))
        ) {
            continue;
        }

        try {
            await processUi5YamlConfig(fs, basePath, ui5Yaml, script);
        } catch (error) {
            logger?.warn(
                `Skipping script '${scriptName}', which refers to the UI5 YAML configuration file '${ui5Yaml}'. ${error.message}`
            );
            continue;
        }

        const { path } = extractUrlDetails(script);
        if (path) {
            await renameSandbox(fs, join(await getWebappPath(basePath), path), logger);
        }
        ensurePreviewMiddlewareDependency(packageJson, fs, packageJsonPath);

        logger?.info(
            `The UI5 YAML configuration file '${ui5Yaml}', has been updated according to script, '${scriptName}'.`
        );
    }
    for (const ui5Yaml of unprocessedUi5YamlFileNames) {
        //at least adjust deprecated preview config of unused ui5 yaml configurations
        try {
            await processUi5YamlConfig(fs, basePath, ui5Yaml, '', true);
        } catch (error) {
            logger?.warn(`Skipping UI5 yaml configuration file '${ui5Yaml}'. ${error.mesage}`);
        }
        logger?.warn(
            `The UI5 YAML configuration file '${ui5Yaml}', is not used in any preview script. Outdated preview middleware will be adjusted, if necessary.`
        );
    }
}
