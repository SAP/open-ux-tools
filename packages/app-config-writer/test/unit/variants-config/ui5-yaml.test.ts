import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';
import type { Editor } from 'mem-fs-editor';
import { join } from 'path';
import { ToolsLogger } from '@sap-ux/logger';
import { updateMiddlewares } from '../../../src/variants-config/ui5-yaml';
import { FileName } from '@sap-ux/project-access';

type YamlFileName = typeof FileName.Ui5MockYaml | typeof FileName.Ui5LocalYaml | typeof FileName.Ui5Yaml;

const middlewareUpdatedMessage = (middleware: 'preview' | 'reload', filename: YamlFileName) =>
    `Updated ${middleware} middleware in ${filename}.`;

const noFileMessage = (filename: YamlFileName) => `Cannot write variants-config to ${filename}. File not existing`;

describe('Test update middleware', () => {
    const logger = new ToolsLogger();
    let fs: Editor;
    const debugLogMock = jest.spyOn(ToolsLogger.prototype, 'debug').mockImplementation(() => {});
    const basePath = join(__dirname, '../../fixtures/variants-config');

    beforeEach(() => {
        jest.clearAllMocks();
        fs = create(createStorage());
    });

    test('add preview middleware config to ui5.yaml file', async () => {
        await updateMiddlewares(fs, basePath, logger);

        expect(fs.read(join(basePath, 'ui5.yaml'))).toMatchSnapshot();
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('preview', FileName.Ui5Yaml));
        expect(debugLogMock).toHaveBeenCalledWith(noFileMessage(FileName.Ui5MockYaml));
        expect(debugLogMock).toHaveBeenCalledWith(noFileMessage(FileName.Ui5LocalYaml));
    });

    test('add preview and reload middleware config to ui5.yaml file', async () => {
        const openSourceConfig = join(basePath, 'open-source-config');
        await updateMiddlewares(fs, openSourceConfig, logger);

        expect(fs.read(join(openSourceConfig, 'ui5.yaml'))).toMatchSnapshot();
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('preview', FileName.Ui5Yaml));
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('reload', FileName.Ui5Yaml));
        expect(debugLogMock).toHaveBeenCalledWith(noFileMessage(FileName.Ui5MockYaml));
        expect(debugLogMock).toHaveBeenCalledWith(noFileMessage(FileName.Ui5LocalYaml));
    });

    test('add preview and reload middleware to local ui5.yaml files', async () => {
        const fioriToolsConfig = join(basePath, 'fiori-tools-config');
        await updateMiddlewares(fs, fioriToolsConfig, logger);

        expect(debugLogMock).toHaveBeenCalledTimes(6);
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('preview', FileName.Ui5Yaml));
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('preview', FileName.Ui5LocalYaml));
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('preview', FileName.Ui5MockYaml));

        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('reload', FileName.Ui5Yaml));
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('reload', FileName.Ui5LocalYaml));
        expect(debugLogMock).toHaveBeenCalledWith(middlewareUpdatedMessage('reload', FileName.Ui5MockYaml));
    });
});
