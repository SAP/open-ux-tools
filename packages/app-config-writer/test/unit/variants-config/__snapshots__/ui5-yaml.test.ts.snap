// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test update middleware add fiori-tools-preview to ui5.yaml file w/o middlewares 1`] = `
"specVersion: \\"3.0\\"
metadata:
  name: fe_lrop_v2
type: application
server:
  customMiddleware:
    - name: fiori-tools-preview
      afterMiddleware: compression
"
`;

exports[`Test update middleware add preview and reload middleware config to ui5.yaml file 1`] = `
"specVersion: \\"3.0\\"
metadata:
  name: fe_lrop_v2
type: application
server:
  customMiddleware:
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        delay: 300
    - name: preview-middleware
      afterMiddleware: fiori-tools-appreload
      configuration:
        flp:
          intent:
            object: hello
            action: world
"
`;

exports[`Test update middleware add preview middleware config to ui5.yaml file 1`] = `
"specVersion: \\"3.0\\"
metadata:
  name: fe_lrop_v2
type: application
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          intent:
            object: hello
            action: world
        rta:
          layer: VENDOR
          editors:
            - path: /editor.html
              developerMode: true
            - path: /my-variants.html
"
`;

exports[`Test update middleware add preview-middleware to ui5.yaml file w/o middlewares 1`] = `
"specVersion: \\"3.0\\"
metadata:
  name: fe_lrop_v2
type: application
server:
  customMiddleware:
    - name: preview-middleware
      afterMiddleware: compression
"
`;
